backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/predicate/impl/ElasticsearchKnnPredicate.java

```diff
@@ -0,0 +1,250 @@
+			}
+			else {
+				// shouldn't really happen as if someone tries this it should fail on `queryElementFactory` lookup.
+				throw new AssertionFailure( "Attempting to use a knn predicate on a non-vector field." );
+			}
+		}
+
+		@Override
+		public void k(int k) {
+			this.k = k;
+		}
+
+		@Override
+		public void vector(Object vector) {
+			if ( !vector.getClass().isArray() ) {
+				throw new IllegalArgumentException( "Vector can only be either a float or a byte array (float[], byte[])." );
+			}
+			if ( !vectorElementsType.equals( vector.getClass().getComponentType() ) ) {
+				throw log.vectorKnnMatchVectorTypeDiffersFromField( absoluteFieldPath, vectorElementsType,
+						vector.getClass().getComponentType() );
+			}
+			if ( Array.getLength( vector ) != indexedVectorsDimension ) {
+				throw log.vectorKnnMatchVectorDimensionDiffersFromField( absoluteFieldPath, indexedVectorsDimension,
+						Array.getLength( vector )
+				);
+			}
+			this.vector = vectorToJsonArray( vector, vectorElementsType );
+		}
+
+		@Override
+		public void filter(SearchPredicate filter) {
+			this.filter = ElasticsearchSearchPredicate.from( scope, filter );
+		}
+
+	}
+
+	private static JsonArray vectorToJsonArray(Object vector, Class<?> vectorElementsType) {
+		// we know it is an array since we've checked it when we got the vector
+		int length = Array.getLength( vector );
+		JsonArray array = new JsonArray( length );
+		for ( int i = 0; i < length; i++ ) {
+			if ( byte.class.equals( vectorElementsType ) ) {
+				array.add( Array.getByte( vector, i ) );
+			}
+			else {
+				array.add( Array.getFloat( vector, i ) );
+			}
+		}
+		return array;
+	}
+
+	private static class ElasticsearchImpl extends ElasticsearchKnnPredicate {
+
+		private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+		private static final JsonAccessor<String> FIELD_ACCESSOR = JsonAccessor.root().property( "field" ).asString();
+		private static final JsonArrayAccessor QUERY_VECTOR_ACCESSOR = JsonAccessor.root().property( "query_vector" ).asArray();
+		private static final JsonAccessor<Integer> K_ACCESSOR = JsonAccessor.root().property( "k" ).asInteger();
+
+		private static final JsonObjectAccessor FILTER_ACCESSOR = JsonAccessor.root().property( "filter" ).asObject();
+		private static final JsonAccessor<Integer> NUM_CANDIDATES_ACCESSOR =
+				JsonAccessor.root().property( "num_candidates" ).asInteger();
+
+
+		private ElasticsearchImpl(Builder<?> builder) {
+			super( builder );
+		}
+
+		@Override
+		public JsonObject toJsonQuery(PredicateRequestContext context) {
+			JsonObject knn = new JsonObject();
+			FIELD_ACCESSOR.set( knn, absoluteFieldPath );
+			K_ACCESSOR.set( knn, k );
+			if ( filter != null ) {
+				FILTER_ACCESSOR.set( knn, filter.toJsonQuery( context ) );
+			}
+			NUM_CANDIDATES_ACCESSOR.set( knn, numberOfCandidates != null ? numberOfCandidates : k );
+			QUERY_VECTOR_ACCESSOR.set( knn, vector );
+
+			context.contributeKnnClause( knn );
```

Comment:
One of the missing things is a check for where knn predicate is added i.e. is it a top-level predicate/should clause of a top-level bool predicate/a clause/filter of some predicate that itself is a clause to some other predicate ...
I don't think we have anything to determine at which level `toJsonQuery` is called,so I was thinking should we add a parent to `PredicateRequestContext` and create  a new context setting current one as a parent any time we call toJsonQuery, so that the context is a tree node and we can decide if the location of a knn predicate is acceptable?

Reply:
>  I don't think we have anything to determine at which level toJsonQuery is called,so I was thinking should we add a parent to PredicateRequestContext and create a new context setting current one as a parent any time we call toJsonQuery, so that the context is a tree node and we can decide if the location of a knn predicate is acceptable?

+1 makes sense, but be careful: we already create new contexts right now, to handle nested contexts. That one could have been modeled as a tree too, but it hasn't. You'll have to make sure things work, and ideally make it a bit more consistent.